File: node/engines/yandex.js
const { getBrowser } = require("../_browser");

async function fetchYandex(query, n = 10) {
  const browser = await getBrowser();
  const page = await browser.newPage();
  try {
    await page.goto(
      `https://yandex.com/images/search?text=${encodeURIComponent(query)}`,
      { waitUntil: "domcontentloaded", timeout: 15000 }
    );
    await page.waitForSelector(".serp-item[data-bem]", { timeout: 10000 }).catch(() => {});
    return await page.$$eval(
      ".serp-item[data-bem]",
      (items, maxN) => {
        const urls = [];
        const seen = new Set();
        for (const item of items) {
          try {
            const bem = JSON.parse(item.getAttribute("data-bem") || "{}");
            const url = bem?.["serp-item"]?.img_href;
            if (url && !seen.has(url)) {
              seen.add(url);
              urls.push(url);
              if (urls.length >= maxN) break;
            }
          } catch {}
        }
        return urls;
      },
      n
    );
  } finally {
    await page.close();
  }
}

module.exports = { fetchYandex };

File: imgsearch_api/engines/yandex.py
import json
from urllib.parse import quote_plus
from .._fetcher import get_fetcher


def fetch_yandex(query: str, n: int = 10) -> list[str]:
    page = get_fetcher().fetch(
        f"https://yandex.com/images/search?text={quote_plus(query)}",
        headless=True,
        timeout=15000,
        wait_selector=".serp-item",
        wait_selector_state="attached",
    )
    urls: list[str] = []
    for item in page.css(".serp-item[data-bem]"):
        try:
            bem = json.loads(item.attrib.get("data-bem", "{}"))
            url = bem.get("serp-item", {}).get("img_href", "")
            if url and url not in urls:
                urls.append(url)
                if len(urls) >= n:
                    break
        except (json.JSONDecodeError, AttributeError):
            continue
    return urls
